src/socket/communicate.rs

```diff
@@ -0,0 +1,239 @@
+    pub struct Listener {
+        /// Bound Unix socket.
+        listener: UnixListener,
+        /// Lock that keeps our socket exclusive.
+        #[allow(dead_code)]
+        bind_lock: BindLock,
+        /// How long to wait for the client to send its
+        /// first message after opening the connection.
+        accept_timeout: Timeout,
+    }
+
+    /// Errors in `accept()`ing a new connection.
+    #[derive(Debug)]
+    pub enum AcceptError {
+        /// something went wrong in the `accept()` syscall.
+        Accept(std::io::Error),
+        /// The client’s message could not be decoded.
+        Message(ReadWriteError),
+    }
+
+    impl Listener {
+        /// Create a new `daemon` by binding to `socket_path`.
+        pub fn new(socket_path: &SocketPath) -> Result<Listener, BindError> {
+            let (l, lock) = socket_path.bind()?;
+            Ok(Listener {
+                listener: l,
+                bind_lock: lock,
+                accept_timeout: DEFAULT_READ_TIMEOUT,
+            })
+        }
+
+        /// Accept a new connection on the socket,
+        /// read the communication type and then delegate to the
+        /// corresponding handling subroutine.
+        ///
+        /// The handler is start in a thread, the thread handle is returned.
+        ///
+        /// This method blocks until a client tries to connect.
+        pub fn accept<F: 'static>(
+            &self,
+            handler: F,
+        ) -> Result<std::thread::JoinHandle<()>, AcceptError>
+        where
+            F: FnOnce(UnixStream, CommunicationType) -> (),
+            F: std::marker::Send,
+        {
+            // - socket accept
+            let (unix_stream, _) = self.listener.accept().map_err(AcceptError::Accept)?;
+            // - read first message as a `CommunicationType`
+            let comm_type: CommunicationType =
+                ReadWriter::<CommunicationType, ConnectionAccepted>::new(&unix_stream)
+                    .react(self.accept_timeout.clone(), |_| ConnectionAccepted())
+                    .map_err(AcceptError::Message)?;
+            // spawn a thread with the accept handler
+            Ok(std::thread::spawn(move || handler(unix_stream, comm_type)))
+        }
+    }
+
+}
+
+/// Clients that can talk to a `Listener`.
+///
+/// `R` is the type of messages this client reads.
+/// `W` is the type of messages this client writes.
+///
+/// The construction of `Client` is only exported for
+/// the pre-defined interactions with the `Listener` we support.
+pub mod client {
+    use super::*;
+    use std::marker::PhantomData;
+
+    /// A `Client` that can talk to a `Listener`.
+    pub struct Client<R, W> {
+        /// Type of interaction with the `Listener`.
+        comm_type: CommunicationType,
+        /// Connected socket.
+        socket: Option<UnixStream>,
+        /// Timeout for reads/writes.
+        timeout: Timeout,
+        phantom_r: PhantomData<R>,
+        phantom_w: PhantomData<W>,
```

Comment:
Maybe should change these `phantom_*` names to indicate why they're there, not that they're phantom. Like to hint that they are used for pinning the read / write data types.